awx/main/tasks.py

```diff
@@ -149,6 +149,81 @@ def handle_setting_changes(self, setting_keys):
             break
 
 
+def _add_remove_queues(app, instance, worker_queues, worker_name):
+    removed_queues = []
+    added_queues = []
+    ig_names = set([ig.name for ig in instance.rampart_groups.all()])
+    worker_queue_names = set([q['name'] for q in worker_queues])
+
+
+    # Remove queues that aren't in the instance group
+    for queue in worker_queues:
+        if queue['name'] in settings.AWX_CELERY_QUEUES_STATIC or \
+                queue['alias'] in settings.AWX_CELERY_QUEUES_STATIC:
+            continue
+
+        if queue['name'] not in ig_names:
+            app.control.cancel_consumer(queue['name'], reply=True, destination=[worker_name])
+            removed_queues.append(queue['name'])
+
+    # Add queues for instance and instance groups
+    for queue_name in ig_names | set([instance.hostname]):
+        if queue_name not in worker_queue_names:
+            app.control.add_consumer(queue_name, reply=True, destination=[worker_name])
+            added_queues.append(queue_name)
+
+    return (added_queues, removed_queues)
+
+
+def _update_celery_routes(instance):
+    logger.debug("Updating celery routes for instance {}".format(instance.hostname))
+
+    tasks = [
+        'awx.main.tasks.cluster_node_heartbeat',
+        'awx.main.tasks.purge_old_stdout_files',
+    ]
+
+    # Instance is, effectively, a controller node
+    if Instance.objects.filter(rampart_groups__controller__instances=instance).exists():
+        tasks.append('awx.main.tasks.awx_isolated_heartbeat')
+
+    for t in tasks:
+        settings.CELERY_ROUTES[t] = {'queue': instance.hostname, 'routing_key': instance.hostname}
+
+
+def _handle_ha_toplogy_changes(celery_worker_name):
+    instance = Instance.objects.me()
+    added_queues = []
+    removed_queues = []
+
+    _update_celery_routes(instance)
+
+    # Try catch this
+    app = Celery(broker=settings.BROKER_URL)
+    celery_host_queues = app.control.inspect([celery_worker_name]).active_queues()
```

Comment:
What kind of data do you typically see for `celery_host_queues`? When I try to run these lines inside the dev container shell I just get `None`, and that's during normal operation.

Reply:
wait, I didn't get that it needed to follow the pattern of `celery@awx`, and it's probably that this was your intent when sending in `request.hostname` from the tasks that called it.